### ISSUE TITLE
mix ci.check error unexpected port exit

### ISSUE THREAD

**Issue by wingyplus** [sasa1977/ci#8]
I found after run `mix ci.check`:

```
$ mix ci.check
Compiling 1 file (.ex)
starting mix compile --warnings-as-errors
CI checks took 0 seconds
** (Mix) mix compile --warnings-as-errors failed: unexpected port exit
```

After investigate, I found that test suite was fail in case run a command with `pty: true`

```
$ mix test
Compiling 1 file (.ex)
Excluding tags: [ci: true]

..........................................................

  1) test GenServer can use pty (OsCmdTest)
     test/os_cmd_test.exs:31
     ** (EXIT from #PID<0.507.0>) %OsCmd.Error{exit_status: nil, message: "echo 1 failed: unexpected port exit", output: nil}

....

Finished in 1.2 seconds
64 tests, 1 failure, 1 excluded

Randomized with seed 950894
```

After I realize that, I modify `check.ex` to set `pty: false` and it works:

```
$ mix ci.check
Compiling 1 file (.ex)
starting mix compile --warnings-as-errors
Compiling 1 file (.ex)
mix compile --warnings-as-errors took 0.6 seconds
starting mix dialyzer
starting mix test
starting mix format --check-formatted
starting mix docs
mix format --check-formatted took 0.5 seconds
Finding suitable PLTs
Checking PLT...
Generating docs...
[:compiler, :eex, :elixir, :kernel, :logger, :mix, :mox, :nimble_parsec, :parent, :stdlib, :telemetry]
Looking up modules in dialyxir_erlang-23.3_elixir-1.11.3_deps-test.plt
Looking up modules in dialyxir_erlang-23.3_elixir-1.11.3.plt
Looking up modules in dialyxir_erlang-23.3.plt
Finding applications for dialyxir_erlang-23.3.plt
Excluding tags: [ci: true]

Finding modules for dialyxir_erlang-23.3.plt
........View "html" docs at "doc/index.html"
...................View "epub" docs at "doc/ci.epub"
mix docs took 1.7 seconds
.Checking 18 modules in dialyxir_erlang-23.3.plt
.

  1) test GenServer can use pty (OsCmdTest)
     test/os_cmd_test.exs:31
     ** (EXIT from #PID<0.414.0>) %OsCmd.Error{exit_status: nil, message: "echo 1 failed: unexpected port exit", output: nil}

.Adding 181 modules to dialyxir_erlang-23.3.plt
................................

Finished in 1.4 seconds
64 tests, 1 failure, 1 excluded

Randomized with seed 422814
mix test took 2.5 seconds
...
```

I try running `os_cmd` command locally it works fine:

```
$ ./priv/os_cmd "-pty=true" "mix" "compile" "--warnings-as-errors"
started)hwoutputmCompiling 9 files (.ex)
"hwoutputmGenerated ci app
$ echo $status
0
```

**Comment by wingyplus** [comment]
Elixir and Erlang on my machine:

```
$ elixir --version
Erlang/OTP 23 [erts-11.2] [source] [64-bit] [smp:8:8] [ds:8:8:10] [async-threads:1] [hipe] [dtrace]

Elixir 1.11.3 (compiled with Erlang/OTP 23)
```

**Comment by sasa1977** [comment]
Yeah, it seems that something blows up when we use `pty` from inside a port, while `pty` works when started from a terminal. The culprit is [this line](https://github.com/sasa1977/ci/blob/develop/ports/os_cmd/program.go#L50) which crashes the go program (with exit code 129).

Since this function is from the [creack/pty library](https://github.com/creack/pty), further debugging is required to understand the issue. Unfortunately I can't say when I will find the time for this.

A quick patch would be to ignore the pty flag on macOS. This is far from perfect, but at least it wouldn't crash the command.

**Comment by iautom8things** [comment]
There looks to be a discussion about this issue in the dependency's repo, that started recently:

https://github.com/creack/pty/issues/127

**Comment by creack** [comment]
I have no idea what a "port" is :) but for context, if you try to start a process in a PTY while not having access to the PTY, then it fails.

As an example, if you try to use `docker run -t` in most CIs, it will fail. Removing the `-t` (i.e. the PTY allocation) fixes the issue. It is not a problem with the PTY lib, but with the caller context.

I don't think creack/pty#127 is related, that one is about reading data *after* the pty closes.  Which is something to be avoided.

**Comment by sasa1977** [comment]
@creack Thanks for the comment!

> I have no idea what a "port" is

For context, from an Elixir (or Erlang) program we start an external go program. The mechanism that support this in Elixir/Erlang world is called a port.

From the go program we then start another external program. Depending on the options passed, that program may be started via `exec` or via `pty`. The relevant go code can be seen [here](https://github.com/sasa1977/ci/blob/bc67646b67255a0df7dff761b1105f8b822e9b5d/ports/os_cmd/program.go#L22-L84), and in particular the pty execution branch is [here](https://github.com/sasa1977/ci/blob/bc67646b67255a0df7dff761b1105f8b822e9b5d/ports/os_cmd/program.go#L49-L61). 

Note that in the pty branch we're checking for [pty.ErrUnsupported](https://github.com/sasa1977/ci/blob/bc67646b67255a0df7dff761b1105f8b822e9b5d/ports/os_cmd/program.go#L52-L54) and in this case fall back to plain execution via `exec`. However, as I mentioned in the [earlier comment](https://github.com/sasa1977/ci/issues/8#issuecomment-834220836), `pty.Start` crashes instead of returning an error, and this is why we get "unexpected port exit", which basically means that the go program has crashed.

It's also worth mentioning that this seems to happen only on macOS (e.g. it works fine on Ubuntu, and the [relevan test](https://github.com/sasa1977/ci/blob/bc67646b67255a0df7dff761b1105f8b822e9b5d/test/os_cmd_test.exs#L31-L33) is passing on GH actions), and (according to the initial report) only when the go program is started via Erlang. So I'm not saying the cause is necessarily in the pty lib, but still I wouldn't expect `pty.Start` to crash.
